src/mcfret.c

```diff
@@ -0,0 +1,366 @@
+	              for (i=0; i<non->singles; i++){
+	                  j=i*non->singles;
+	              /* Propagate vector*/
+	                  if (non->propagation==1) propagate_vec_coupling_S(non,Hamil_i_e,vecr+j,veci+j,non->ts,1);
+	                      if (non->propagation==0){
+	                          if (non->thres==0 || non->thres>1){
+	                              propagate_vec_DIA(non,Hamil_i_e,vecr+j,veci+j,1);
+	                          } 
+                        else {
+	                          elements=propagate_vec_DIA_S(non,Hamil_i_e,vecr+j,veci+j,1);
+	                          if (samples==non->begin){
+	                              if (t1==0){
+		                                if (x==0){
+		                                    printf("Sparce matrix efficiency: %f pct.\n",(1-(1.0*elements/(non->singles*non->singles)))*100);
+		                                    printf("Pressent tuncation %f.\n",non->thres/((non->deltat*icm2ifs*twoPi/non->ts)*(non->deltat*icm2ifs*twoPi/non->ts)));
+		                                    printf("Suggested truncation %f.\n",0.001);
+		                                }
+	                              }
+	                          }
+	                        }
+	                      }
+                  }
+            
+              }/*We are closing the loop over time delays -t1 times*/
+         } /*We are closing the cluster loop*/
+
+         /*Create NISE log file: Why not immediately in the begining?*/ 
+         log=fopen("NISE.log","a");
+         fprintf(log,"Finished sample %d\n",samples);
+          
+         time_now=log_time(time_now,log);
+         fclose(log);
+    }/*Closing the loop over samples*/
+    
+    /*The calculation is finished, lets write output*/
+    log=fopen("NISE.log","a");
+    fprintf(log,"Finished Calculating Response!\n");
+    fprintf(log,"Writing to file!\n");  
+    fclose(log);
+
+    if (non->cluster!=-1){
+        printf("Of %d samples %d belonged to cluster %d.\n",samples,Ncl,non->cluster);
+        if (samples==0){ /*Avoid dividing by zero*/ 
+            samples=1;
+        }
+    }
+
+    fclose(H_traj);
+    if (non->cluster!=-1){
+        fclose(Cfile);
+    }
+
+  /* Save time domain response */
+    absorption_matrix=fopen("TD_absorption_matrix.dat","w");
+    emission_matrix=fopen("TD_emission_matrix.dat","w");
+    for (t1=0;t1<non->tmax1;t1+=non->dt1){
+        fprintf(absorption_matrix,"%f %e %e\n",t1*non->deltat,re_S_Abs[t1]/samples,im_S_Abs[t1]/samples);
+        fprintf(emission_matrix, "%f %e %e\n",t1*non->deltat,re_S_Emi[t1]/samples,im_S_Emi[t1]/samples);
+    }
+    fclose(absorption_matrix);
+    fclose(emission_matrix);
+    
+    
+    /*Free the memory*/
+    
+    free(re_S_Abs);
+    free(im_S_Abs);
+    free(re_S_Emi);
+    free(im_S_Emi);
+
+    
+    free(vecr);	
+    free(veci);
+
+    free(dens_mat);
+
+    
+}
+void density_matrix(float *density_matrix, float *Hamiltonian_i,t_non *non){
```

Comment:
It is quite confusing that the `density_matrix()` function has an input called `density_matrix`. Maybe `create_density_matrix()` is a better function name.